File: windows/src/GoalMaker.Core/Planning/ReminderSchedule.cs
namespace GoalMaker.Core.Planning;

/// <summary>
/// What a device does with its reminders when it looks (docs/reminders.md, pinned by
/// contracts/vectors/reminders.json). Only one timer is armed at a time: when it goes off the device
/// shows what arrived since its last look and arms the <see cref="Next"/> one, so a device that was
/// asleep catches up and a reminder is shown once.
/// </summary>
public static class ReminderSchedule
{
    /// <summary>Every reminder that has a time, soonest first, ties settled by id so both devices agree.</summary>
    public static IReadOnlyList<ScheduledReminder> Resolve(
        IEnumerable<ReminderItem> reminders,
        IReadOnlyDictionary<string, TaskItem> tasks,
        QuietHours quietHours)
    {
        var scheduled = new List<ScheduledReminder>();
        foreach (var reminder in reminders)
        {
            if (tasks.TryGetValue(reminder.TaskId, out var task) && ReminderRules.Fires(reminder, task, quietHours) is { } at)
            {
                scheduled.Add(new ScheduledReminder(reminder.Id, task.Id, task.Title, at, reminder.Important));
            }
        }

        return [.. scheduled.OrderBy(item => item.At).ThenBy(item => item.Id, StringComparer.Ordinal)];
    }

    /// <summary>What to show now: the reminders that arrived after <paramref name="since"/>, the last look, up to <paramref name="now"/>.</summary>
    public static IReadOnlyList<ScheduledReminder> Due(
        IEnumerable<ReminderItem> reminders,
        IReadOnlyDictionary<string, TaskItem> tasks,
        QuietHours quietHours,
        DateTime since,
        DateTime now) =>
        [.. Resolve(reminders, tasks, quietHours).Where(item => item.At > since && item.At <= now)];

    /// <summary>The reminder to arm the next timer for, or null when nothing is waiting.</summary>
    public static ScheduledReminder? Next(
        IEnumerable<ReminderItem> reminders,
        IReadOnlyDictionary<string, TaskItem> tasks,
        QuietHours quietHours,
        DateTime now) =>
        Resolve(reminders, tasks, quietHours).FirstOrDefault(item => item.At > now);

    /// <summary>
    /// Which of the notifications on screen (<paramref name="shown"/>, by reminder id) have to go: their
    /// reminder was handled, snoozed, moved later or deleted, or its task finished, here or on the other device.
    /// </summary>
    public static IReadOnlyList<string> Stale(
        IEnumerable<string> shown,
        IEnumerable<ReminderItem> reminders,
        IReadOnlyDictionary<string, TaskItem> tasks,
        DateTime now)
    {
        var byId = reminders.ToDictionary(reminder => reminder.Id, StringComparer.Ordinal);
        return [.. shown.Where(id => !StillDue(id)).Order(StringComparer.Ordinal)];

        bool StillDue(string id) =>
            byId.TryGetValue(id, out var reminder)
            && tasks.TryGetValue(reminder.TaskId, out var task)
            && ReminderRules.Due(reminder, task) is { } due
            && due <= now;
    }
}

File: android/app/src/main/kotlin/com/goalmaker/app/application/planning/ReminderSchedule.kt
package com.goalmaker.app.application.planning

import com.goalmaker.app.domain.planning.QuietHours
import java.time.LocalDateTime

/**
 * What a device does with its reminders when it looks (docs/reminders.md, pinned by
 * contracts/vectors/reminders.json). Only one alarm is armed at a time: when it goes off the device
 * shows what arrived since its last look and arms the [next] one, so a device that was off catches
 * up, a reminder is shown once, and no alarm quota is spent on reminders far ahead.
 */
object ReminderSchedule {
    /** Every reminder that has a time, soonest first, ties settled by id so both devices agree. */
    fun resolve(
        reminders: List<ReminderItem>,
        tasks: Map<String, TaskItem>,
        quietHours: QuietHours,
    ): List<ScheduledReminder> = reminders
        .mapNotNull { reminder ->
            val task = tasks[reminder.taskId] ?: return@mapNotNull null
            val at = ReminderRules.fires(reminder, task, quietHours) ?: return@mapNotNull null
            ScheduledReminder(reminder.id, task.id, task.title, at, reminder.important)
        }
        .sortedWith(compareBy({ it.at }, { it.id }))

    /** What to show now: the reminders that arrived after [since], the last look, up to [now]. */
    fun due(
        reminders: List<ReminderItem>,
        tasks: Map<String, TaskItem>,
        quietHours: QuietHours,
        since: LocalDateTime,
        now: LocalDateTime,
    ): List<ScheduledReminder> = resolve(reminders, tasks, quietHours).filter { it.at.isAfter(since) && !it.at.isAfter(now) }

    /** The reminder to arm the next alarm for, or null when nothing is waiting. */
    fun next(
        reminders: List<ReminderItem>,
        tasks: Map<String, TaskItem>,
        quietHours: QuietHours,
        now: LocalDateTime,
    ): ScheduledReminder? = resolve(reminders, tasks, quietHours).firstOrNull { it.at.isAfter(now) }

    /**
     * Which of the notifications on screen ([shown], by reminder id) have to go: their reminder was
     * handled, snoozed, moved later or deleted, or its task finished, here or on the other device.
     */
    fun stale(
        shown: Collection<String>,
        reminders: List<ReminderItem>,
        tasks: Map<String, TaskItem>,
        now: LocalDateTime,
    ): List<String> {
        val byId = reminders.associateBy(ReminderItem::id)
        return shown.filter { id ->
            val reminder = byId[id] ?: return@filter true
            val task = tasks[reminder.taskId] ?: return@filter true
            val due = ReminderRules.due(reminder, task)
            due == null || due.isAfter(now)
        }.sorted()
    }
}
